node/core/approval-voting/src/lib.rs

```diff
@@ -1802,39 +1915,43 @@ where
 			)),
 	};
 
-	let (claimed_core_index, assigned_candidate_hash) =
-		match block_entry.candidate(candidate_index as usize) {
-			Some((c, h)) => (*c, *h),
+	// The Compact VRF modulo assignment cert has multiple core assignments.
+	let mut backing_groups = Vec::new();
+	let mut claimed_core_indices = Vec::new();
+	let mut assigned_candidate_hashes = Vec::new();
+
+	for candidate_index in candidate_indices.iter_ones() {
```

Comment:
The `CheckAndImportAssignment` message should either
1. specify that the candidate indicates must have a sane length (i.e. within n_cores for the session)
2. check that it does.

The approval-distribution subsystem needs to check this anyway, so I'm in favor of (1) but I also expect that (2) is cheap, given that we've already loaded the session info.

Reply:
Likewise with the `CoreBitfield` in the assignment.

Reply:
I think we shouldn't need to fetch session info in `approval distribution`, I know it will await only for first query per relay parent + on new session, but the 2 extra cache lookups for all approval messages will add overhead to all messages before any book keeping.

I propose to add a hardcoded max size in `approval distribution`, for example 500 should be reasonable assuming we're gonna be doing `Execution(Core) Groups` by the time we get to this usage/scale. But in `approval voting` we do (2) as we already have the session info. If `approval voting` finds the candidate indices or core indices aren't within n_cores, we return the error to `approval-distribution` which disconnects the peer.

WDYT ?

Reply:
done